File: src/podcast/services/encoder.py
import asyncio
import logging
import os
import shutil
import time
import uuid

from pydub import AudioSegment
from sqlalchemy import func, select

from podcast.config import settings
from podcast.database import get_session
from podcast.models import Episode

logger = logging.getLogger(__name__)


def _encode(episode_id: uuid.UUID, audio_dir: str) -> tuple[str, int, int]:
    """Synchronous MP3 encoding — returns (filename, duration_seconds, file_size)."""
    wav_path = os.path.join(audio_dir, f"{episode_id}.wav")
    if not os.path.exists(wav_path):
        raise FileNotFoundError(f"WAV file not found: {wav_path}")

    mp3_filename = f"{episode_id}.mp3"
    mp3_path = os.path.join(audio_dir, mp3_filename)

    logger.info("Encoding MP3 for episode %s", episode_id)

    # Load WAV, convert to mono, export as MP3
    audio = AudioSegment.from_wav(wav_path)
    audio = audio.set_channels(1)  # Mono
    audio = audio.set_frame_rate(44100)
    audio.export(mp3_path, format="mp3", bitrate="128k")

    duration_seconds = int(len(audio) / 1000)
    file_size = os.path.getsize(mp3_path)

    # Clean up WAV and segments
    os.remove(wav_path)
    segments_dir = os.path.join(audio_dir, "segments", str(episode_id))
    if os.path.isdir(segments_dir):
        shutil.rmtree(segments_dir)

    return mp3_filename, duration_seconds, file_size


async def encode_mp3(episode_id: uuid.UUID) -> dict:
    """Convert WAV to MP3 and update episode metadata. Returns metrics dict."""
    # Look up user_id to determine audio directory
    async with get_session() as db:
        episode = await db.get(Episode, episode_id)
        if not episode:
            raise ValueError(f"Episode {episode_id} not found")
        user_id = episode.user_id

    user_audio_dir = os.path.join(settings.audio_dir, str(user_id))

    # Run encoding in a thread
    t0 = time.monotonic()
    mp3_filename, duration_seconds, file_size = await asyncio.to_thread(
        _encode, episode_id, user_audio_dir
    )
    encode_duration = time.monotonic() - t0

    # Update DB with results
    async with get_session() as db:
        episode = await db.get(Episode, episode_id)
        if not episode:
            raise ValueError(f"Episode {episode_id} not found")

        # Get next episode number (scoped per user)
        result = await db.execute(
            select(func.coalesce(func.max(Episode.episode_number), 0))
            .where(Episode.user_id == user_id)
        )
        next_number = result.scalar_one() + 1

        episode.audio_filename = mp3_filename
        episode.audio_duration_seconds = duration_seconds
        episode.audio_size_bytes = file_size
        episode.episode_number = next_number

    logger.info(
        "Encoding complete for episode %s: %ds, %d bytes, episode #%d (%.1fs)",
        episode_id,
        duration_seconds,
        file_size,
        next_number,
        encode_duration,
    )
    return {
        "duration_seconds": round(encode_duration, 2),
        "audio_duration_seconds": duration_seconds,
        "output_size_bytes": file_size,
    }

File: src/services/encoder.ts
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import fs from "node:fs";
import path from "node:path";
import { eq, sql } from "drizzle-orm";
import { db } from "../database.js";
import { episodes } from "../schema.js";
import { settings } from "../config.js";
import { createLogger } from "../log-handler.js";

const execFileAsync = promisify(execFile);
const log = createLogger("encoder");

async function encode(
  episodeId: string,
  audioDir: string
): Promise<{ filename: string; durationSeconds: number; fileSize: number }> {
  const wavPath = path.join(audioDir, `${episodeId}.wav`);
  if (!fs.existsSync(wavPath)) {
    throw new Error(`WAV file not found: ${wavPath}`);
  }

  const mp3Filename = `${episodeId}.mp3`;
  const mp3Path = path.join(audioDir, mp3Filename);

  log.info("Encoding MP3 for episode %s", episodeId);

  // Convert to mono 44.1kHz 128kbps MP3
  await execFileAsync("ffmpeg", [
    "-i", wavPath,
    "-ac", "1",
    "-ar", "44100",
    "-b:a", "128k",
    "-y",
    mp3Path,
  ]);

  // Get duration via ffprobe
  const { stdout } = await execFileAsync("ffprobe", [
    "-v", "error",
    "-show_entries", "format=duration",
    "-of", "default=noprint_wrappers=1:nokey=1",
    mp3Path,
  ]);
  const durationSeconds = Math.floor(parseFloat(stdout.trim()));
  const fileSize = fs.statSync(mp3Path).size;

  // Clean up WAV and segments
  fs.unlinkSync(wavPath);
  const segmentsDir = path.join(audioDir, "segments", episodeId);
  if (fs.existsSync(segmentsDir)) {
    fs.rmSync(segmentsDir, { recursive: true });
  }

  return { filename: mp3Filename, durationSeconds, fileSize };
}

export async function encodeMp3(
  episodeId: string
): Promise<Record<string, unknown>> {
  // Look up user_id
  const [episode] = await db
    .select()
    .from(episodes)
    .where(eq(episodes.id, episodeId))
    .limit(1);
  if (!episode) throw new Error(`Episode ${episodeId} not found`);

  const userAudioDir = path.join(settings.audioDir, episode.userId);

  const t0 = performance.now();
  const { filename, durationSeconds, fileSize } = await encode(
    episodeId,
    userAudioDir
  );
  const encodeDuration = (performance.now() - t0) / 1000;

  // Assign episode number and update atomically (advisory lock prevents races)
  const nextNumber = await db.transaction(async (tx) => {
    await tx.execute(
      sql`SELECT pg_advisory_xact_lock(hashtext(${episode.userId}))`
    );
    const [{ maxNum }] = await tx
      .select({
        maxNum: sql<number>`coalesce(max(${episodes.episodeNumber}), 0)`,
      })
      .from(episodes)
      .where(eq(episodes.userId, episode.userId));
    const next = (maxNum ?? 0) + 1;

    await tx
      .update(episodes)
      .set({
        audioFilename: filename,
        audioDurationSeconds: durationSeconds,
        audioSizeBytes: fileSize,
        episodeNumber: next,
      })
      .where(eq(episodes.id, episodeId));

    return next;
  });

  log.info(
    "Encoding complete for episode %s: %ds, %d bytes, episode #%d (%.1fs)",
    episodeId,
    durationSeconds,
    fileSize,
    nextNumber,
    encodeDuration
  );

  return {
    duration_seconds: Math.round(encodeDuration * 100) / 100,
    audio_duration_seconds: durationSeconds,
    output_size_bytes: fileSize,
  };
}
